Common/Securities/Option/StrategyMatcher/OptionStrategyMatcher.cs

```diff
@@ -0,0 +1,181 @@
+        {
+            _marketPriceProvider = marketPriceProvider;
+        }
+
+        public IEnumerable<OptionPosition> Enumerate(OptionPositionCollection positions)
+        {
+            if (positions.IsEmpty)
+            {
+                yield break;
+            }
+
+            var marketPrice = _marketPriceProvider(positions.Underlying);
+
+            var longPositions = new List<OptionPosition>();
+            var shortPuts = new SortedDictionary<decimal, OptionPosition>();
+            var shortCalls = new SortedDictionary<decimal, OptionPosition>();
+            foreach (var position in positions)
+            {
+                if (!position.Symbol.HasUnderlying)
+                {
+                    yield return position;
+                }
+
+                if (position.Quantity > 0)
+                {
+                    longPositions.Add(position);
+                }
+                else
+                {
+                    switch (position.Right)
+                    {
+                        case OptionRight.Put:
+                            shortPuts.Add(position.Strike, position);
+                            break;
+
+                        case OptionRight.Call:
+                            shortCalls.Add(position.Strike, position);
+                            break;
+
+                        default:
+                            throw new ApplicationException(
+                                "The skies are falling, the oceans rising - you're having a bad time"
+                            );
+                    }
+                }
+            }
+        }
+    }
+
+    public class OptionStrategyMatcher
+    {
+        public OptionStrategyMatcherOptions Options { get; }
+
+        public OptionStrategyMatcher(OptionStrategyMatcherOptions options)
+        {
+            Options = options;
+        }
+
+        //public OptionStrategyMatch FindBestMatch(OptionPositionCollection positions)
+        //{
+
+        //}
+
+        public IEnumerable<OptionStrategyMatch> Match(OptionPositionCollection positions)
+        {
+            foreach (var definition in Options.Definitions)
+            {
+
+            }
+
+            yield break;
+        }
+
+        public IReadOnlyList<OptionStrategyMatch> MatchOnce(OptionPositionCollection positions)
+        {
+            var matches = new List<OptionStrategyMatch>();
+            foreach (var definition in Options.Definitions)
+            {
+                while (true)
+                {
+                    using (var enumerator = definition.Match(positions).GetEnumerator())
+                    {
+                        if (!enumerator.MoveNext())
+                        {
+                            break;
+                        }
+
+                        var match = enumerator.Current;
+                        positions = positions.Accept(match);
+
+                    }
+                }
+            }
+
+            return null;
```

Comment:
Looks like this guy is still WIP

Reply:
Yep -- that was my bad -- I held back the good stuff :P